### PR TITLE
Endpoint to receive notarization results in BAPI

### PR DESCRIPTION
**Pull request by pabloaguayo**
### What does this Pull Request do?
The Buyer API receives the notarization result and starts processing it

### Are there points in the code the reviewer needs to double check?
Mutability of notarization results. Structural validation of sellers.

### Why was this Pull Request needed?
To be protocol v2 compliant.

### Does this Pull Request meet the acceptance criteria?
- [x] I have read the [Contribution guidelines](https://github.com/wibsonorg/buyer-sdk/CONTRIBUTING.md).
- [x] I have read the [Code of Conduct](https://github.com/wibsonorg/buyer-sdk/CODE_OF_CONDUCT.md).
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: buyer-api/src/utils/routes.js [hunk 1]
```diff
@@ -28,4 +28,32 @@ const validateAddress = paramName => async (req, res, next) => {
   return undefined;
 };
 
-export { errorHandler, asyncError, validateAddress };
+/**
+ * Middleware to check if an notarization result is valid or not.
+ * @public
+ */
+const validateFields = () => (req, res, next) => {
```
Comment (nicoayala) [inline buyer-api/src/utils/routes.js]: There is no need to have a factory function here. You could just export `export const validateFields = (req, res, next) => {...}` and use it in the router

File: buyer-api/src/utils/routes.js [hunk 2]
```diff
@@ -28,4 +28,32 @@ const validateAddress = paramName => async (req, res, next) => {
...
+  const notarizationResult = req.body;
+
+  if (!notarizationResult.orderId) {
+    res.status(422).json({
```
Comment (nicoayala) [inline buyer-api/src/utils/routes.js]: We are using `express-boom` to standarize error responses, so I would response with smth like:
`res.boom.badData('Validation error', { errors: ['Field "orderId" is missing or malformed'] });`

File: buyer-api/test/operations/notarizationResultReception.test.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import {
+  someNotarizationRequest,
+  someNotarizationResult,
+  someNotarizationResultWithDuplicatedAddresses,
+  someNotarizationResultWithNonRequestedAddresses,
+} from './notarizationResultReception.fixture';
+
+const it = test.serial;
+
+it('call notarizationResultReception', async (assert) => {
```
Comment (Meketh) [inline buyer-api/test/operations/notarizationResultReception.test.js]: Test description should be about what this test is trying to assert.
(And by convention are written on simple present)

File: buyer-api/test/operations/notarizationResultReception.test.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  notarizationResultReception(
+    someNotarizationRequest,
+    someNotarizationResult,
+  );
+  assert.snapshot(addTransactionJob.lastCall.args);
+});
+
+it('call notarizationResultReception with not requested addresses', async (assert) => {
```
Comment (Meketh) [inline buyer-api/test/operations/notarizationResultReception.test.js]:
```suggestion
it('filters not requested addresses', async (assert) => {
```

File: buyer-api/test/operations/notarizationResultReception.test.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  notarizationResultReception(
+    someNotarizationRequest,
+    someNotarizationResultWithNonRequestedAddresses,
+  );
+  assert.is(addTransactionJob.lastCall.lastArg.sellers.length, 2);
+});
+
+it('call notarizationResultReception with duplicated addresses', async (assert) => {
```
Comment (Meketh) [inline buyer-api/test/operations/notarizationResultReception.test.js]:
```suggestion
it('filters duplicated addresses', async (assert) => {
```

File: buyer-api/src/facades/index.js [hunk 6]
```diff
@@ -10,6 +10,7 @@ export {
   getNotaryInfo,
   getNotariesInfo,
   fetchAndCacheNotary,
+  getNotarizationRequest,
```
[the remark this answers is not in the document]
Comment (nicoayala) [reply buyer-api/src/facades/index.js]: @Meketh @pabloaguayo Don't worry for this now, I will fix it when I start the Request Notarization Process

File: buyer-api/src/operations/notarizationResultReception.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+/**
+ * take notarization result, validate and enqueue for transfer
+ * WARNING keep in mind that notarizationResult will be updated with the validated sellers
+ * @param {NotarizationRequest} notarizationRequest original request sent for notarization
+ * @param {NotarizationResult} notarizationResult results related with request done by notary
+ */
+export function notarizationResultReception(notarizationRequest, notarizationResult) {
+  // validate input
```
[the remark this answers is not in the document]
Comment (genuinefafa) [reply buyer-api/src/operations/notarizationResultReception.js]: what's the schema you are referring? can we have an example?

File: buyer-api/src/operations/notarizationResultReception.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+ * @property {NumberLike} orderId - tracking Id for the order
+ * @property {String} callbackUrl - callback url for the notarization result
+ * @property {SellersInNotarizationRequest[]} sellers - list of sellers
+ */
+
+/**
+ * take notarization result, validate and enqueue for transfer
+ * WARNING keep in mind that notarizationResult will be updated with the validated sellers
+ * @param {NotarizationRequest} notarizationRequest original request sent for notarization
+ * @param {NotarizationResult} notarizationResult results related with request done by notary
```
[the remark this answers is not in the document]
Comment (genuinefafa) [reply buyer-api/src/operations/notarizationResultReception.js]: `NotarizationResult` is an object, @nicoayala was going to provide it, is that so?

[the remark this answers is not in the document]
Comment (genuinefafa) [reply buyer-api/src/operations/notarizationResultReception.js]: `NotarizationResult` is an object, @nicoayala was going to provide it, is that so?

File: buyer-api/src/routes/notarizationResult.js [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+
+const router = express.Router();
+
+/**
+ * @swagger
+ * /notarization-result/:notarizationRequestId:
+ *   post:
+ *     parameters:
+ *       - name: notarizationRequestId
```
[the remark this answers is not in the document]
Comment (genuinefafa) [reply buyer-api/src/routes/notarizationResult.js]: a json example, you mean @Meketh ? i think that should go into the docs or wiki, not the code.

File: buyer-api/test/operations/notarizationResultReception.mock.js [hunk 10]
```diff
@@ -0,0 +1,10 @@
+import td from 'testdouble';
+import sinon from 'sinon';
+import test from 'ava';
+
+export const notarizationRequests = { put: sinon.spy() };
```
[the remark this answers is not in the document]
Comment (genuinefafa) [reply buyer-api/test/operations/notarizationResultReception.mock.js]: but we are not using that method, why should we mock something not needed?

File: buyer-api/src/routes/notarizationResult.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (Meketh) [reply buyer-api/src/routes/notarizationResult.js]: Swagger docs are retrieved by swagger-api.
So this is the actual documentation, thats why it's important to keep endpoints code simple.
See [src/routes/dataOrders/createDataOrder.js](https://github.com/wibsonorg/buyer-sdk/blob/b57e5c5deb2832a82f4f3f793469914725eed981/buyer-api/src/routes/dataOrders/createDataOrder.js)

File: buyer-api/test/operations/notarizationResultReception.mock.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (Meketh) [reply buyer-api/test/operations/notarizationResultReception.mock.js]: What's the difference between `getNotarizationRequest` and `notarizationRequests.get`.
Does it have some kind of extra logic ? Is it just a wrapper ? Is it even necessary ?
Why should we abstract a call to the store in another layer ?
Shouldn't it just be a call to the store ? Or do it needs some kind of pre processing ?
And if so, why don't we just store data already processed ?

File: buyer-api/src/operations/notarizationResultReception.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (nicoayala) [reply buyer-api/src/operations/notarizationResultReception.js]: Just to be clear, we'll include input validation as a middleware when we find some time, and yes we can use `src/schema.js` for that purpose.

File: buyer-api/src/facades/index.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (genuinefafa) [reply buyer-api/src/facades/index.js]: ok! where shall we leave this TODO @nicoayala ?

[the remark this answers is not in the document]
Comment (nicoayala) [reply buyer-api/src/facades/index.js]: Sure!

File: buyer-api/test/operations/notarizationResultReception.fixture.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+  id: 94,
+  address: '0x2d419c641351e0baa7f54328ecabf58c5e4a56f1',
+  result: 'not_audited',
+  decryptionKeyEncryptedWithMasterKey: '0x912f8f484454e3a38f7535fbf6b7f0035a0fe27c028163348965eb9369fcca8c',
+});
+
+export const someNotarizationResultWithDuplicatedAddresses = { ...someNotarizationResult };
+someNotarizationResultWithDuplicatedAddresses.sellers =
+  someNotarizationResultWithDuplicatedAddresses.sellers.slice(0);
+someNotarizationResultWithDuplicatedAddresses.sellers.push({
```
[the remark this answers is not in the document]
Comment (genuinefafa) [reply buyer-api/test/operations/notarizationResultReception.fixture.js]: see new version

File: buyer-api/test/operations/notarizationResultReception.mock.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (genuinefafa) [reply buyer-api/test/operations/notarizationResultReception.mock.js]: already done

**Comment by genuinefafa** [comment]
@nicoayala can you validate latest changes so we can move forward to develop?

**Pull request merged**